
# Feature-level QC 

**Feature QC is a critical yet often overlooked component in single-cell analysis, essential for accurate interpretation of cell functions and reliable cross-sample comparisons. Inadequate feature QC might result in the retention of low-quality data caused by dropout events, poor cell suspension, or ambient RNA contamination.**
In this report, we only support **Feature Metrics** to overview the quality of features (genes/proteins) in the dataset. And the part of **Systematic background noise** can be analyzed by the **`SingleCellMQC`** R package.

## Feature summary {.tabset }

```{r feature_qc_overview, echo=FALSE, results='asis', warning=FALSE, message=FALSE }
  generate_primary_card(
  header_content = "**Key components overview: **",
    card_title =NULL,
  body_content = "
- **Marker PCT:** (If available markers) The detection rates of cell markers across different celltypes.
- **Metrics plot:** The distribution characteristics of features across cells within each sample.
  "
  )
```


```{r Feature marker PCT, echo=FALSE, results='asis', message=FALSE, warning=FALSE}

if(length(data$Feature$Section1$markerPCT)!=0){
  cat(paste0("### ", "Marker PCT", " {.tabset .unnumbered}\n"))
  cat("This table displays the detection rates (percentage of cells with expression >0) of classical cell markers across different cell populations, serving as a quality control metric for both gene detection and sample quality assessment. By examining the detection percentage of markers in their expected cell types, we can evaluate potential dropout rates and overall data quality. This quality control approach is particularly valuable for single-cell RNA sequencing data. \n")
    cat('\n\n')

  subchunkify(data$Feature$Section1$markerPCT,seed="markerPCT")

  cat('\n\n')
}
  

```



```{r Feature information, echo=FALSE, results='asis', message=FALSE, warning=FALSE}

if(length(data$Feature$Section1$base)!=0 ){
  cat(paste0("### ", "Metrics plot", " {.tabset .unnumbered .tabset-dropdown}\n"))
  cat('\n\n')

  base = data$Feature$Section1$base
  
  
  
  for (i in base) {
    cat(paste0("#### ", i, " {.tabset .unnumbered}\n"))
    if("RNA"%in%  data$Feature$Section1$assay ){
      cat(paste0("##### ", "RNA", " {.tabset .unnumbered}\n"))
      cat("Bar plots (5% bins) showing the distribution of gene detection rates in RNA data. Corresponding the right bar plots highlight the top 10 genes with the highest detection rates. Mean-variance plot for RNA: The x-axis represents the mean of log-normalized counts, and the y-axis represents the variance of log-normalized counts for each gene. Each blue dot denotes a single gene. 
          \n\n")
      cat("![](", paste0(cell_dir, "/Feature_QC/FeatureMetrics/",  paste0(i,"_RNA.png")), ")")
      cat('\n\n')
    }
    if("ADT"%in%  data$Feature$Section1$assay ){
      cat(paste0("##### ", "ADT", " {.tabset .unnumbered}\n"))
      cat("Bar plots (5% bins) showing the distribution of protein detection rates in RNA data. Corresponding the right bar plots highlight the top 10 protein with the highest detection rates. Mean-variance plot for ADT: The x-axis represents the mean of CLR (centered log-ratio) transformed counts, and the y-axis represents the variance of CLR transformed counts for each antibody-derived tag (ADT). Each blue dot denotes a single protein ADT. 
          \n\n")
      cat("![](", paste0(cell_dir, "/Feature_QC/FeatureMetrics/",  paste0(i,"_ADT.png")), ")")
      cat('\n\n')
    }
      cat('\n\n')
  }
}

```


