# Sample-level QC 

**The main tasks of sample-level QC include: (i) Sample quality assessment: detecting technical artifacts that may be caused by library construction, insufficient sequencing depth, amplification biases, or contaminations; (ii) Outlier sample detection: identifying samples with unexpected cell type compositions; and (iii) Sample identity verification: identifying incorrect sample labeling caused by errors in sample handling.**


------------------------------------------------------------------------

## QC summary

```{r summary, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if(length(data$Sample$Section1$Metrics )!=0){
  # cat( paste0(data$Sample$Section1$Metrics$note, collapse = "  \n") )
  note_grey(paste0(data$Sample$Section1$Metrics$note, collapse = "  \n"))
  cat('\n\n')
    generate_primary_card(
  header_content = "**Key QC tasks summary information: **",
    card_title =NULL,
  body_content = data$Sample$summary)
  cat('\n\n')
}

```

## QC details {.tabset }

```{r Key components overview, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
  generate_primary_card(
  header_content = "**Key components overview: **",
    card_title =NULL,
  body_content = "
- **10X Genomics Metrics (if available):** From Cell Ranger output. 
- **RNA & ADT Metrics:** From SingleCellMQC, including RNA, ADT metrics and QC results.
- **V(D)J Metrics:** From SingleCellMQC, including VDJ metrics and QC results.
- **Celltype% outlier sample detection:** To identify deviations in cell type compositions.
- **Predicted sex information:** To avoid potential recording mistakes and data transmission errors.
  "
  )
 

```

```{r cellranger warning, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if (!is.null(data$Sample$Section1$Metrics$Metrics_10X$base) && 
    length(data$Sample$Section1$Metrics$Metrics_10X$base) > 0){
  
  cat("
### 10X Genomics metrics {.tabset .unnumbered}

This section summarizes 10X metrics alerts generated from [10X Genomics CellRanger](https://github.com/10XGenomics/cellranger) standards. Alerts are categorized as **errors** (indicating critical sample issues) or **warnings** (highlighting potential concerns), as detailed below:
      ")
  cat('\n\n')
}


if(length(data$Sample$Section1$Alert )!=0){
  if(length(data$Sample$Section1$Alert$table$Error)!=0){
    message <- paste0(paste0("- ", length(data$Sample$Section1$Alert$table$Error)," Error Samples: ", paste0(data$Sample$Section1$Alert$table$Error, collapse = ", "), " \n"),
    paste0( "- ", length(data$Sample$Section1$Alert$table$Warning)," Warning Samples: ", paste0(data$Sample$Section1$Alert$table$Warning, collapse = ", "), ""),
    " \n"
    )
    warning_alert_orange(message)
    
  }
  subchunkify(data$Sample$Section1$Alert$interactive_table,seed="cellranger_alerts")
  cat("\n\n")
}

# Check and output 10X metrics if data exists
if (!is.null(data$Sample$Section1$Metrics$Metrics_10X$base) && 
    length(data$Sample$Section1$Metrics$Metrics_10X$base) > 0) {
  subchunkify(data$Sample$Section1$Metrics$Metrics_10X$base, seed="Metrics_10X")
  cat('\n\n')
}


```



### RNA & ADT metrics {.tabset .unnumbered}

```{r metrics outlier, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if(length(data$Sample$Section1$outlier_metrics)!=0){
  if(length(data$Sample$Section1$outlier_metrics$table)!=0){
    table_data <- data$Sample$Section1$outlier_metrics$table
    cat_word <- do.call(c, lapply(names(table_data), function(x){
      paste0( "- ", length(table_data[[x]])," ", x," outlier samples: ", paste0(table_data[[x]],  collapse = ", "), " \n")
    }))
    warning_alert( paste0("Metrics outlier results (MAD Statistics): \n\n ",paste0(cat_word, collapse = ""),   " \n",
         "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"))
    
  }
  
  generate_primary_card(
  header_content = NULL,
    card_title =NULL,
  body_content = data$Sample$Section1$outlier_metrics$interactive_table,
  footer_content= "For the table, `type`: define the direction of outlier detection (`lower` for values below the median threshold or `higher` for values above it); `log`: whether metrics were log-transformed prior to analysis. ")
  
  cat("\n\n")

}

```

```{r metrics information, echo=FALSE, results='asis', warning=FALSE, message=FALSE}

if(length(data$Sample$Section1$Metrics )!=0){

# Check and output Basic information if data exists
if (!is.null(data$Sample$Section1$Metrics$count$base) && 
    length(data$Sample$Section1$Metrics$count$base) > 0) {
  
  cat(paste0("#### Basic information", " {.tabset .unnumbered}\n"))
  subchunkify(data$Sample$Section1$Metrics$count$base, seed="count base information")
  cat('\n\n')
}

if (!is.null(data$Sample$Section1$Metrics$count$statistics) && 
    length(data$Sample$Section1$Metrics$count$statistics) > 0) {
  subchunkify(data$Sample$Section1$Metrics$count$statistics, seed="count base information statistics")
  cat('\n\n')
}

# Check and output Statistical summary if data exists
stat_data <- data$Sample$Section1$Metrics$summary
if (!is.null(stat_data) && length(stat_data) > 0) {
  cat(paste0("#### Descriptive statistics of cell QC metrics per sample", " {.tabset .unnumbered}\n"))
  
# subchunkify(metrics_help_details(
#   title = "Abbreviations",
#   items = c(
#     "Estimated number of cells",
#     "Median UMI per cell"
#   )
# ), seed=paste0("tag_"))
  
cat("\n\n")

  i=1000
  for (g in names(stat_data)) {
    if (!is.null(stat_data[[g]]$base) && length(stat_data[[g]]$base) > 0) {
      cat('##### ', g, "{.unnumbered}",' \n')
      subchunkify(stat_data[[g]]$base, seed=paste0("summary_",g))
      cat("\n\n")
    }
  }
}

    
    cat("\n\n")
  }


```


```{r vdj information, echo=FALSE, results='asis', warning=FALSE, message=FALSE}
if(length(data$Sample$Section1$VDJ )!=0){
  
  cat("
### V(D)J { .tabset .unnumbered}
")
  cat('\n\n')
  cat("This section summarizes the V(D)J repertoire analysis, including V(D)J chain composition, V(D)J subtype composition, V(D)J gene usage (geneUsage), clonotype quantification (clonalQuant), and CDR3 sequence length profiles (CDR3). Gene usage frequencies (geneUsage) and clonal expansion metrics (clonalQuant) are derived using the [scRepertoire](https://github.com/ncborcherding/scRepertoire) package, which also processes CDR3 length distributions (CDR3). Detailed results are presented below. The details below: \n\n")
  stat_data <- data$Sample$Section1$VDJ
  i=1000
  
  generate_primary_card(
  header_content = "**Notes**",
    card_title =NULL,
  body_content = gsub("> -", "", paste0(data$Sample$Section1$VDJ$note, collapse = "  \n"))
)
  
  # note_grey( paste0(data$Sample$Section1$VDJ$note, collapse = "  \n") ) 
  
  cat('\n\n')
  
  for (g in c("V(D)J chain", "V(D)J subtype") ) {
      cat('#### ', g, "{.unnumbered}",  ' \n')
      subchunkify(stat_data[[g]]$base,seed=paste0("VDJ_",g))
      cat("\n\n")
      subchunkify(stat_data[[g]]$statistics,seed=paste0("VDJ_statistics_",g))
      cat("\n\n")
  }
  cat("\n\n")

}


##geneUsage
if( length(data$Sample$Section1$VDJ$geneUsage )!=0 ){
  vdj_data <- data$Sample$Section1$VDJ$geneUsage
  name <- names(vdj_data)
  cat(paste0("#### geneUsage", " {.tabset .unnumbered}\n"))

  for (i in name) {
    cat(paste0("##### ", i, " {.tabset .unnumbered}\n"))
    subchunkify(vdj_data[[i]],seed= paste0(i, "_geneUsage") )
    cat('\n\n')
  }
  cat('\n\n')
}


##clonalQuant
if( length(data$Sample$Section1$VDJ$clonalQuant )!=0 ){
  vdj_data <- data$Sample$Section1$VDJ$clonalQuant
  name <- names(vdj_data)
  cat(paste0("#### clonalQuant", " {.tabset .unnumbered}\n"))

  for (i in name) {
    cat(paste0("##### ", i, " {.tabset .unnumbered}\n"))
    subchunkify(vdj_data[[i]],seed= paste0(i, "_clonalQuant") )
    cat('\n\n')
  }
  cat('\n\n')
}


##CDR3
if( length(data$Sample$Section1$VDJ$CDR3 )!=0 ){
  vdj_data <- data$Sample$Section1$VDJ$CDR3
  name <- names(vdj_data)
  cat(paste0("#### CDR3", " {.tabset .unnumbered}\n"))

  for (i in name) {
    cat(paste0("##### ", i, " {.tabset .unnumbered}\n"))
    cat("![](", paste0(cell_dir, "/Sample_QC/VDJ/CDR3/",  paste0(i,"_CDR3.png")), ")")
    cat("<p align='middle' style='font-size: 16px; color: black;'>**Figure: ** The distribution of CDR3 length. </p> \n\n")
        cat(paste0( "**IF you want to see interactive plotly object for details, you can see it by click the link ", "[plotly_link](", paste0( "./SingleCellMQC/plot/Sample_QC/VDJ/CDR3/",  paste0(i, "_CDR3.html")), ")", ".**\n\n")) 
    cat('\n\n')
  }

  cat('\n\n')
}



```


```{r cell type composition outlier, echo=FALSE, results='asis', warning=FALSE, message=FALSE}


if(length(data$Sample$Section2$outlier_comp)!=0){
  
  cat("
### Celltype% outlier sample detection {.tabset .unnumbered}

**Outlier sample detection in SingleCellMQC focuses on identifying deviations in cell type compositions. ** Deviations in cell type composition within samples from the same experimental group are often indicative of potential technical issues, such as prolonged processing time, variations in storage conditions, experimental processing errors, and batch effects, all of which can compromise the integrity of the data. To address this issue, SingleCellMQC incorporates cell type composition analysis as a additional outlier detection approach. This section include two main approaches: 1) **Reference-based outlier analysis**: Identification of outliers via deviation from reference cell type proportions; 2) **PCA-based outlier analysis**: Detection of inter-sample outliers via aberrant cell type proportionds within cohort. The details below:
")
  cat('\n\n')
  comp_data <- data$Sample$Section2$outlier_comp
  
  ###warining message
  
  
  if(length(comp_data$com_table)!=0){
      table_data <- comp_data$com_table
      
      # subset freq >= 3
      freq_table <- as.data.frame(table( do.call(c, table_data) ))
      freq_table <- freq_table[freq_table$Freq >=3,]
      freq_table <- freq_table[ order(freq_table$Freq, decreasing =T) , ]
      cat_freq_word3ormore <- paste0( "- ", length(freq_table[,1]), " warning samples (celltype% outliers >= 3 ): ", paste0(freq_table[,1], collapse = ", ")," \n")
      
      # cat_word <- do.call(c, lapply(names(table_data), function(x){
      #   paste0( "- ", length(table_data[[x]])," ", x," outlier samples: ", paste0(table_data[[x]],  collapse = ", "), " \n")
      # }))
      # warning_alert_orange( paste0("Reference-based outlier analysis: \n\n ", cat_freq_word3ormore, " \n",
      #      "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"))
      warning_alert(
        text =  paste0(
          cat_freq_word3ormore,
          " \n",
          "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.",
          "\n"
        ),
        title = "Warning! (Reference-based analysis)"
      )
      
      
  }
  
  
  if(length(comp_data$name_ellipse)!=0){
    ellipse_pa <- paste0( "PCA-based outlier analysis (Ellipse): \n\n ",
    paste0("- ", "", length(comp_data$name_ellipse)," cell type composition outlier samples:  ", paste0(comp_data$name_ellipse, collapse = ", "), ""),
    " \n"
    )
  }else{
    ellipse_pa <- NULL
  }
  
  if(length(comp_data$name_dbscan)!=0){
    DBSCAN_pa <- paste0( "PCA-based outlier analysis (DBSCAN): \n\n ",
    paste0("- ", "", length(comp_data$name_dbscan)," cell type composition outlier samples: ", paste0(comp_data$name_dbscan, collapse = ", "), ""),
    " \n"
    )
  }else{
    DBSCAN_pa <- NULL
  }
  
  if(!is.null(ellipse_pa) | !is.null(DBSCAN_pa) ){
    
    warning_alert(
      text =  paste0(ellipse_pa,  " \n", DBSCAN_pa,  " \n", 
                                "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"),
      title = "Warning! (PCA-based analysis)"
    )
    
    # warning_alert_orange(paste0(ellipse_pa,  " \n", DBSCAN_pa,  " \n", 
    #                             "**Notice: ** The above prediction results represent samples that may have potential abnormalities, and are for reference only. Please interpret the results in combination with the actual situation of the data.", "\n"))
  }
  
  
  generate_primary_card(
    header_content = NULL,
    card_title = NULL,
    body_content = comp_data$CellTypePCT,
    footer_content = "Note: The per-sample cell type annotation by [ScType](https://github.com/IanevskiAleksandr/sc-type) method, and then quantifies cell type proportions per sample."
  )
    
    # subchunkify(comp_data$CellTypePCT,seed="CellTypePCT")
    cat("\n\n")
    # cat("Note: The per-sample cell type annotation by [ScType](https://github.com/IanevskiAleksandr/sc-type) method, and then quantifies cell type proportions per sample")
    # cat("\n\n")

    cat(paste0("#### Reference-based outlier analysis", " {.tabset .unnumbered}\n"))
    cat("SingleCellMQC established the reference ranges for nine major cell types across ten human tissues by analyzing 2,299 scRNA-seq samples from the Deeply Integrated Human Single-Cell Omics data [(DISCO)](https://www.immunesinglecell.org/) database. Samples with deviations from these tissue-specific reference cell type compositions are automatically flagged as outlier samples using the `FindCommonPCTOutlier` function. \n")
  cat('\n\n')

    if(length(comp_data$com_table)!=0){
      subchunkify(comp_data$com_interactive_table,seed="com_interactive_table")
    }
    cat("\n\n")
    
    
    cat("\n\n")
    cat(paste0("#### PCA-based outlier analysis", " {.tabset .unnumbered}\n"))
    
    cat("A PCA-based outlier detection method using confidence ellipse or Density-Based Spatial Clustering of Applications with Noise (DBSCAN) to identify samples with unusual cell type proportions within sample groups was also provided in this section. Additionally, contributions of each cell type composition to PCs were calculated, providing further insights into potential outliers and ensuring the reliability of the dataset for downstream analysis. \n")
    cat("\n\n")
 
  subchunkify(comp_data$interactive_table,seed="outlier_comp_interactive_table")
  cat("\n\n")

  cat(paste0("##### Ellipse", " {.tabset .unnumbered} \n\n"))
  subchunkify(comp_data$pca_ellipse, seed="outlier_comp_ellipse", fig_width=8)
  cat("\n\n")
  cat("PCA of cell proportions across samples. The Ellipse panel highlights the outlier identification based on the 95% confidence ellipse, where the blue circle represents the range of the 95% confidence ellipse.")
  cat("\n\n")

  cat(paste0("##### DBSCAN", " {.tabset .unnumbered} \n\n"))
  subchunkify(comp_data$pca_dbscan, seed="outlier_comp_dbscan", fig_width=8)
  cat("PCA of cell proportions across samples. The outliers were identified by DBSCAN clustering. The blue ellipse indicates the 95% confidence region of the core population.")
  cat("\n\n")
}

```




```{r Sample label verification, echo=FALSE, results='asis', warning=FALSE}

if(!is.null(data$Sample$Section3)){
  
  cat('### Predicted sex information {.unnumbered}

To avoid potential recording mistakes and data transmission errors, checking whether the file labeling is consistent with analyzed data features is necessary at the beginning of sample QC. In this section, by default, sex-specific genes, including XIST, DDX3Y, UTY, and RPS4Y1, can be evaluated to verify the alignment with recorded sex information.')
  cat('\n\n')

  generate_primary_card(
    header_content = "**The predicted sex information**",
    card_title = NULL,
    body_content = data$Sample$Section3$identity,
    footer_content = NULL
  )
  
  
# identity <- data$Sample$Section3$identity
# subchunkify(reactablefmtr::add_source(identity, "Table: The predicted sex information"), seed="identity")
  generate_primary_card(
    header_content = "**Notes**",
    card_title = NULL,
    body_content = paste0("- **XIST** is female-expressed from its function in X-chromosome inactivation. **DDX3Y**, **UTY** and **RPS4Y1** are Y-chromosome-specific. **`Predict`** column represents the analyzed sex.  \n", '-  The analyzed sex results: **Female: ', sum(data$Sample$Section3$num %in% "Female") , " ;  Male: ", sum(data$Sample$Section3$num %in% "Male"), " ;**"),
    footer_content = NULL
  )

 # note_grey(paste0("> **NOTE:**  \n>  \n> - **XIST** is female-expressed from its function in X-chromosome inactivation. **DDX3Y**, **UTY** and **RPS4Y1** are Y-chromosome-specific. **`Predict`** column represents the analyzed sex.  \n", '> -  The analyzed sex results: **Female: ', sum(data$Sample$Section3$num %in% "Female") , " ;  Male: ", sum(data$Sample$Section3$num %in% "Male"), " ;**"))
cat("\n\n")
}


```




